Step4 - 요금 조회 기능 PR

KangJunJun opened the pull request:
잘 부탁 드립니다..!

KangJunJun committed:
refactor: Step3 피드백 반영

- RequestMapping 활용
- favorite을 member가 아닌 favoriteRepository로 조회하도록 변경
- favorite domain 객체에 dto 의존관계 제거
- RestControllerAdvice 활용

KangJunJun committed:
docs: Step4 - 요구사항 리스트업

KangJunJun committed:
feat: 거리별 요금 정책 추가

KangJunJun committed:
feat: 노선별 요금 정책 기능 추가

KangJunJun committed:
feat: 나이별 요금 정책 기능 추가

limwoobin commented on src/main/java/nextstep/subway/exception/ExceptionAdvice.java:
```diff
@@ -0,0 +1,27 @@
+    public ResponseEntity handleIllegalArgsException(AuthorizationException e) {
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+    }
+}
```
advice를 잘 적용해주셨네요 ㅎㅎ
다만 RuntimeException은 예상치 못한 예외에서 발생할것같은데 badRequest로 응답하는게 맞을까 싶습니다
보다 예외를 나누고 해당 예외에 맞는 status code를 응답하는게 클라이언트입장에서
어떤 이슈인지 파악하기에 더 용이해보입니다
(추가로 예외문구도 있으면 좋겠네요!)

limwoobin commented on src/main/java/nextstep/subway/fare/domain/DiscountFarePolicy.java:
```diff
@@ -0,0 +1,40 @@
+    over19(19,999,0),
+    over13_under18(13, 18, 0.2),
+    over6_under12(6, 12, 0.5),
+    under5(1,5,1);
```
enum은 상수이니만큼 대문자로 선언해주시는게 좋습니다
추가로 나이에 대한 요금계산이니만큼 클래스명으로 명시해주는게 좋을것같아요!

limwoobin commented on src/main/java/nextstep/subway/fare/domain/DiscountFarePolicy.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Arrays;
+
+public enum DiscountFarePolicy {
```
enum으로 처리해주신거 좋습니다!

limwoobin commented on src/main/java/nextstep/subway/fare/domain/Fare.java:
```diff
@@ -0,0 +1,61 @@
+import javax.persistence.Embeddable;
+
+@Embeddable
+public class Fare {
```
총 지불요금인데 Line에서 직접적으로 객체참조로 쓰이고있네요
추가요금인지, 그냥 지하철이용에따른 지불요금인지 책임이 명확하지 않은데
둘을 분리하는건 어떨까요??

@Embedded로 선언하신만큼 해당 객체에서 명확한 책임을 가지고있는게 좋을것같아요 ㅎㅎ

limwoobin commented on src/main/java/nextstep/subway/favorite/ui/FavoriteController.java:
```diff
@@ -11,33 +11,29 @@
 import org.springframework.web.bind.annotation.*;
 
 @RestController
+@RequestMapping("/favorites")
```
피드백 반영 좋습니다!

limwoobin commented on src/main/java/nextstep/subway/line/domain/Line.java:
```diff
@@ -20,6 +21,9 @@ public class Line extends BaseEntity {
     @Embedded
     private Sections sections = new Sections();
 
+    @Embedded
+    private Fare extraFare;
```
앞서 말씀드린 부분에 대해 고민해주시면 좋을듯합니다!

limwoobin commented on src/main/java/nextstep/subway/line/domain/Lines.java:
```diff
@@ -0,0 +1,37 @@
+            graph.setEdgeWeight(edge, section.getDistance());
+            edge.addSection(section);
+        }
+    }
```
Lines는 노선목록인데 여기서 해당 경로를 등록하는건 책임에 맞지 않아보이네요
PathFinder에서 처리하되 List<Line>을 그냥 PathFinder로 넘기는건 어떨까요?

limwoobin commented on src/main/java/nextstep/subway/path/domain/Path.java:
```diff
@@ -22,4 +29,21 @@ public List<Long> getStationIds() {
+                .map(Line::getExtraFare)
+                .max(Comparator.comparing(Fare::value))
+                .orElseThrow(() -> new IllegalArgumentException("추가요금을 찾을 수 없습니다"));
+    }
```
경로 객체에서 요금계산을 하는게 역할에 맞진 않을것같습니다..
FareCalculator 와 같은 객체로 분리하는건 어떨까요?

limwoobin commented on src/main/java/nextstep/subway/path/domain/Path.java:
```diff
@@ -22,4 +29,21 @@ public List<Long> getStationIds() {
+                .map(SectionEdge::getLine)
+                .map(Line::getExtraFare)
+                .max(Comparator.comparing(Fare::value))
+                .orElseThrow(() -> new IllegalArgumentException("추가요금을 찾을 수 없습니다"));
```
추가로 IllegalArgumentException 를 선언해주셧는데 advice에선 해당 예외를
캐치하고 있지는 않네요 ㅎㅎ

limwoobin commented on src/main/java/nextstep/subway/path/domain/SectionEdge.java:
```diff
@@ -0,0 +1,17 @@
+import nextstep.subway.line.domain.Section;
+import org.jgrapht.graph.DefaultWeightedEdge;
+
+public class SectionEdge extends DefaultWeightedEdge {
+    private Section section;
+
+    public void addSection(Section section) {
+        this.section = section;
+    }
+
+    public Line getLine() {
+        return section.getLine();
+    }
+}
```
오 SectionEdge를 만들어주셧군요 ㅎㅎ

limwoobin commented on src/main/java/nextstep/subway/path/dto/PathResponse.java:
```diff
@@ -1,22 +1,27 @@
-        return new PathResponse(stationsResponse(stations), distance);
+    public static PathResponse from(List<Station> stations, Path path, LoginMember loginMember ) {
+
+        return new PathResponse(stationsResponse(stations), path.getDistance(), path.calculate(loginMember.getAge()).value());
```
response객체에서는 경로, 거리, 요금만 받고
해당 계산은 서비스로직이나 바깥에서 처리하는게어떨까요??

limwoobin commented on src/test/java/nextstep/subway/fare/domain/DiscountFarePolicyTest.java:
```diff
@@ -0,0 +1,35 @@
+import org.junit.jupiter.api.Test;
+
+@DisplayName("할인 정책 단위테스트")
+public class DiscountFarePolicyTest {
```
단위테스트에 public은 제거해주셔도 좋겟네요

limwoobin commented on src/test/java/nextstep/subway/fare/domain/DiscountFarePolicyTest.java:
```diff
@@ -0,0 +1,35 @@
+@DisplayName("할인 정책 단위테스트")
+public class DiscountFarePolicyTest {
+    
+    @DisplayName("청소년인 경우 운임에서 350원을 공제한 금액의 20% 할인된다.")
+    @Test
+    void 청소년_할인_정책() {
+        Fare 청소년 = DiscountFarePolicy.calculate(Fare.from(1350), 18);
+        Fare 성인 = DiscountFarePolicy.calculate(Fare.from(1350), 19);
+        assertThat(청소년.value()).isEqualTo(1150);
+        assertThat(성인.value()).isEqualTo(1350);
+    }
+
+    @DisplayName("어린이인 경우 운임에서 350원을 공제한 금액의 50% 할인된다.")
+    @Test
+    void 어린이_할인_정책() {
+        Fare 어린이 = DiscountFarePolicy.calculate(Fare.from(1350), 12);
+        Fare 청소년 = DiscountFarePolicy.calculate(Fare.from(1350), 13);
+        assertThat(어린이.value()).isEqualTo(850);
+        assertThat(청소년.value()).isEqualTo(1150);
+    }
+
+    @DisplayName("1세보다 작은 나이는 유효하지 않습니다.")
+    @Test
+    void 나이가_0세인_경우() {
+        assertThatIllegalArgumentException().isThrownBy(() -> DiscountFarePolicy.calculate(Fare.from(1350), 0));
+    }
```
가능하면 모든 케이스에 대해 검증하시는게 좋을것같아요 ㅎㅎ
성인에 대해서도 넣어주심 좋겟네요!

limwoobin commented on src/test/java/nextstep/subway/fare/domain/FareTest.java:
```diff
@@ -0,0 +1,32 @@
+import org.junit.jupiter.api.Test;
+
+@DisplayName("지하철 요금 도메인 단위테스트")
+public class FareTest {
```
거리, 나이를 모두 고려한 테스트케이스도 있으면 좋겠네요!

limwoobin commented on src/test/java/nextstep/subway/path/acceptance/PathAcceptanceTest.java:
```diff
@@ -81,6 +83,7 @@ void findPath() {
 
         // then
         assertThat(response.as(PathResponse.class).getDistance()).isEqualTo(5);
+        assertThat(response.as(PathResponse.class).getFare()).isEqualTo(1250);
```
인수테스트에도 거리, 나이에 대한 검증이 있으면 더 신뢰도가 높지 않을까 싶습니다!

limwoobin replied:
토큰도 같이 넣어서 회원/비회원 모두 동작하는지 확인이 필요할것같아요!

limwoobin reviewed (approved):
approve 잘못입력

limwoobin reviewed (changes requested):
안녕하세요 준호님
4단계 미션 잘 진행해주셧습니다!

몇가지 고민해보면 좋을부분이 있어 코멘트남겨드렸습니다
확인하시고 다시 요청부탁드립니다!

KangJunJun committed:
refactor: step4 피드백 반영

- enum 대문자로 변경
- 계산로직을 FareCalculator 클래스로 분리
- exception 클래스 분리 및 적용
- finder 기능은 pathFinder 클래스에게 위임
- response 의 계산 로직을 서비스쪽으로 이동

limwoobin commented on src/main/java/nextstep/subway/exception/ExceptionAdvice.java:
```diff
@@ -0,0 +1,25 @@
+//    @ExceptionHandler(RuntimeException.class)
+//    public ResponseEntity handleIllegalArgsException(RuntimeException e) {
+//        return ResponseEntity.badRequest().build();
+//    }
```
주석제거를 깜빡하신것같아요 ㅎㅎ

limwoobin commented on src/main/java/nextstep/subway/exception/SubwayException.java:
```diff
@@ -0,0 +1,21 @@
+package nextstep.subway.exception;
+
+import org.springframework.http.HttpStatus;
+
+public class SubwayException extends RuntimeException {
+    private final HttpStatus status;
+    private final String message;
+
+    public SubwayException(HttpStatus status, String message) {
+        this.status = status;
+        this.message = message;
+    }
+
+    public HttpStatus getStatus() {
+        return status;
+    }
+
+    public String getMessage() {
+        return message;
+    }
+}
```
오 커스텀 Exception에서 메시지와 status를 같이 받으시는군요
굉장히 좋습니다 ㅎㅎ

limwoobin commented on src/main/java/nextstep/subway/line/application/LineService.java:
```diff
@@ -39,7 +41,7 @@ public List<LineResponse> findLines() {
 
     public Line findLineById(Long id) {
-        return lineRepository.findById(id).orElseThrow(RuntimeException::new);
+        return lineRepository.findById(id).orElseThrow(() -> new SubwayException(HttpStatus.NOT_FOUND,"라인을 찾을 수 없습니다"));
```
커스텀 예외, 문구 적용해주신거 좋네요 ㅎㅎ
액션별 줄개행도 고려해주심이 좋겠습니다!

limwoobin commented on src/main/java/nextstep/subway/auth/ui/AuthenticationPrincipalArgumentResolver.java:
```diff
@@ -26,6 +30,11 @@ public boolean supportsParameter(MethodParameter parameter) {
     public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
         String credentials = AuthorizationExtractor.extract(webRequest.getNativeRequest(HttpServletRequest.class));
+        if (credentials == null) {
+            return new LoginMember();
```
토큰이없는 경우에 대해서는 비회원 전용 객체를 만드는것은 어떨까요??

limwoobin reviewed (approved):
안녕하세요 준호님
4단계 피드백도 잘 반영해주셨네요

몇가지 코멘트를 남겨드렸지만 이번 미션에서 의도한 요구사항은 모두 충족되었다고 생각되어
머지하도록 하겠습니다.
정말 고생많으셨습니다!

limwoobin commented:
피드백이 반영되지 않은 부분이 몇군데 보였는데요
미션이 종료되더라도 한번 고민해보면서 다시 작성해보는것을 추천드립니다 ㅎㅎ
이후 논의하고싶은 부분이나 고민되는 부분들에 대해서는 dm주셔도 좋습니다!
